source/math/FloatConvexHull2D.ooc

```diff
@@ -21,55 +21,126 @@ import FloatTransform2D
 import text/StringTokenizer
 
 FloatConvexHull2D: class {
-	// Points must be in clockwise order
-	// TODO: Implement computation of convex hull from a set of points
 	_points: VectorList<FloatPoint2D>
+	_hull: VectorList<FloatPoint2D> = null
 	points ::= this _points
+	hull ::= this _hull
 	count ::= this _points count
-	init: func ~fromPoints (=_points) // Note: Does not compute the convex hull, only assigns it
+	init: func ~fromPoints (=_points, computeHull := false) { //TODO: Compute the convex hull per default
+		// Note, if hull is not computed, points must be in clockwise order
+		if (computeHull)
+			this computeHull()
+		else {
+			this _hull = VectorList<FloatPoint2D> new(this _points count)
+			this _hull = this _points copy()
+		}
+	}
 	init: func ~fromBox (box: FloatBox2D) {
 		this _points = VectorList<FloatPoint2D> new(4)
+		this _hull = VectorList<FloatPoint2D> new(4)
 		this _points add(box leftTop)
 		this _points add(box leftBottom)
 		this _points add(box rightBottom)
 		this _points add(box rightTop)
+		this _hull = this _points copy()
+	}
+	free: override func {
+		this _points free()
+		this _hull free()
+		super()
 	}
 	contains: func ~Point (point: FloatPoint2D) -> Bool {
 		result := true
-		for (i in 0 .. this count - 1)
-			if (This _isOnLeft(this _points[i], this _points[i + 1], point)) {
+		for (i in 0 .. this _hull count - 1)
+			if (result && This _isOnLeft(this _hull[i], this _hull[i + 1], point))
 				result = false
-				break
-			}
-		if (This _isOnLeft(this _points[this count - 1], this _points[0], point))
+		if (This _isOnLeft(this _hull[this _hull count - 1], this _hull[0], point))
```

Comment:
The other things I mentioned are still a performance problem, of course - I just think `%` is a small expense compared to those things.